auction_keeper/main.py

```diff
@@ -307,64 +317,131 @@ def check_auction(self, id: int):
             # Try to remove the auction so the model terminates and we stop tracking it.
             # If auction has already been removed, nothing happens.
             self.auctions.remove_auction(id)
+            return False
 
         # Check if the auction is finished.
         # If it is finished and we are the winner, `deal` the auction.
         # If it is finished and we aren't the winner, there is no point in carrying on with this auction.
         elif auction_finished:
             if input.guy == self.our_address:
-                # TODO this should happen asynchronously
-
                 # Always using default gas price for `deal`
-                self.strategy.deal(id).transact(gas_price=DefaultGasPrice())
+                self._run_future(self.strategy.deal(id).transact_async(gas_price=DefaultGasPrice()))
+
+                # Upon winning a flip or flop auction, we may need to replenish Dai to the Vat.
+                # Upon winning a flap auction, we may want to withdraw won Dai from the Vat.
+                self.rebalance_dai()
 
             else:
                 # Try to remove the auction so the model terminates and we stop tracking it.
                 # If auction has already been removed, nothing happens.
                 self.auctions.remove_auction(id)
+            return False
 
         else:
-            auction = self.auctions.get_auction(id)
+            return True
 
-            # Feed the model with current state
-            auction.feed_model(input)
+    def feed_model(self, id: int):
+        assert isinstance(id, int)
 
-            output = auction.model_output()
-            if output is not None:
-                bid_price, bid_transact = self.strategy.bid(id, output.price)
+        auction = self.auctions.get_auction(id)
+        input = self.strategy.get_input(id)
 
-                if bid_price is not None and bid_transact is not None:
-                    # if no transaction in progress, send a new one
-                    transaction_in_progress = auction.transaction_in_progress()
-                    if transaction_in_progress is None:
-                        self.logger.info(f"Sending new bid @{output.price} (gas_price={output.gas_price})")
+        # Feed the model with current state
+        auction.feed_model(input)
 
-                        auction.price = bid_price
-                        auction.gas_price = UpdatableGasPrice(output.gas_price)
-                        auction.register_transaction(bid_transact)
+    def handle_bid(self, id: int, auction: Auction):
+        assert isinstance(id, int)
+        assert isinstance(auction, Auction)
+
+        output = auction.model_output()
+        # self.logger.debug(f'model output for auction {id} is {output}')
```

Comment:
Same as previous comment ^^ -- not a huge fan of having log statements commented out and pushing them. Lets either include it as debug or remove it